pvlib/tests/ivtools/test_sdm.py

```diff
@@ -0,0 +1,377 @@
+    result_expected = {'I_L_ref': 9.45232,
+                       'I_o_ref': 3.22460e-10,
+                       'R_s': 0.297814,
+                       'R_sh_ref': 125.798,
+                       'a_ref': 1.59128,
+                       'alpha_sc': 0.005658,
+                       'EgRef': 1.121,
+                       'dEgdT': -0.0002677,
+                       'irrad_ref': 1000,
+                       'temp_ref': 25}
+    assert np.allclose(pd.Series(result), pd.Series(result_expected),
+                       rtol=1e-4)
+
+
+@requires_scipy
+def test_fit_desoto_failure():
+    with pytest.raises(RuntimeError) as exc:
+        sdm.fit_desoto(v_mp=31.0, i_mp=8.71, v_oc=38.3, i_sc=9.43,
+                       alpha_sc=0.005658, beta_voc=-0.13788,
+                       cells_in_series=10)
+    assert ('Parameter estimation failed') in str(exc.value)
+
+
+@requires_scipy
+@requires_statsmodels
+def test_fit_desoto_sandia(cec_params_cansol_cs5p_220p):
+    # this test computes a set of IV curves for the input fixture, fits
+    # the De Soto model to the calculated IV curves, and compares the fitted
+    # parameters to the starting values
+    params = cec_params_cansol_cs5p_220p['params']
+    params.pop('Adjust')
+    specs = cec_params_cansol_cs5p_220p['specs']
+    effective_irradiance = np.array([400., 500., 600., 700., 800., 900.,
+                                     1000.])
+    temp_cell = np.array([15., 25., 35., 45.])
+    ee = np.tile(effective_irradiance, len(temp_cell))
+    tc = np.repeat(temp_cell, len(effective_irradiance))
+    iph, io, rs, rsh, nnsvth = pvsystem.calcparams_desoto(
+        ee, tc, alpha_sc=specs['alpha_sc'], **params)
+    sim_ivcurves = pvsystem.singlediode(iph, io, rs, rsh, nnsvth, 300)
+    sim_ivcurves['ee'] = ee
+    sim_ivcurves['tc'] = tc
+
+    result = sdm.fit_desoto_sandia(sim_ivcurves, specs)
+    modeled = pd.Series(index=params.keys(), data=np.nan)
+    modeled['a_ref'] = result['a_ref']
+    modeled['I_L_ref'] = result['I_L_ref']
+    modeled['I_o_ref'] = result['I_o_ref']
+    modeled['R_s'] = result['R_s']
+    modeled['R_sh_ref'] = result['R_sh_ref']
+    expected = pd.Series(params)
+    assert np.allclose(modeled[params.keys()].values,
+                       expected[params.keys()].values, rtol=5e-2)
+
+
+@requires_scipy
+@requires_statsmodels
+def test_fit_pvsyst_sandia(npts=3000):
+    spec_list = ['cells_in_series', 'alpha_sc', 'beta_voc', 'descr']
+    iv_specs = dict.fromkeys(spec_list)
+    keylist = ['i_sc', 'i_mp', 'v_mp', 'v_oc', 'poa', 'tc', 'ee']
+    ivcurves = dict.fromkeys(keylist)
+
+    with open(os.path.join(BASEDIR, 'PVsyst_demo.txt'), 'r') as f:
+
+        Ns, aIsc, bVoc, descr = f.readline().split(',')
+
+        iv_specs.update(
+            cells_in_series=int(Ns), alpha_sc=float(aIsc),
+            beta_voc=float(bVoc), descr=descr)
+
+        strN, strM = f.readline().split(',')
+        N = int(strN)
+        M = int(strM)
+
+        isc = np.empty(N)
```

Comment:
generally inefficient to initialize an empty array and assign values element by element but fine to ignore if the test is fast (< 1 s)

Reply:
Agree, the format of the file being read doesn't help (a few header rows, then for each IV curve: one line of 7 floats, a block of numbers for the voltage array, a second block of numbers for the current array, length of voltage and current arrays can be different for different IV curves).

I doubt the test is <1s, but the major time sink is the function being tested `sdm.fit_pvyst_sandia`.